### PR TITLE
BUG: usecols kwarg accepts string when it should only allow list-like or callable.

### LINKED ISSUE #20529: Unhelpful error message when loading a single column with `read_csv` and `usecols`
#### Code
```python
>>> import pandas as pd
>>> df = pd.DataFrame({'x': [0,1], 'x1': [2,3]})
>>> df.to_csv('tmp.csv', index=False)
>>> pd.read_csv('tmp.csv', usecols='x')
   x
0  0
1  1
>>> pd.read_csv('tmp.csv', usecols=['x1'])
   x1
0   2
1   3
>>> pd.read_csv('tmp.csv', usecols='x1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/matt/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py", line 709, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/matt/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py", line 449, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/home/matt/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py", line 818, in __init__
    self._make_engine(self.engine)
  File "/home/matt/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py", line 1049, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/home/matt/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py", line 1740, in __init__
    raise ValueError("Usecols do not match names.")
ValueError: Usecols do not match names.
```

#### Problem description
When using `usecols` to load a single column, one needs to have either a single-character column name or provide an array-like object.  In the example above, `pd.read_csv('tmp.csv', usecols='x')` and `pd.read_csv('tmp.csv', usecols=['x1'])` work as expected; however, things break down for `pd.read_csv('tmp.csv', usecols='x1')`.  The corresponding error message `ValueError: Usecols do not match names.` is not very helpful either. 

#### Expected Output
It would be nice if there were some type checking done on `usecols` so that things don't break in the example above.   At the least, the error message should be a bit more helpful; e.g., `ValueError: Usecols should be array-like.`

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-37-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.7.1
patsy: None
dateutil: 2.7.1
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by minggli**
- [x] closes #20529 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Since string is iterable, when passed into usecols in read_csv or read_table, it is currently treated as array of characters instead of being caught properly. For example, when usecols='bar', it is interpreted as ['b', 'a' ,'r'] in TextReader, and raising a column not found error in _validate_usecols_names when it really should raise ValueError as invalid value from _validate_usecols_arg, before the param being passed into TextReader.

It's a bug in _validate_usecols_arg and TextReader which lack handling of string-type iterable.

```
>>> import os
>>> from itertools import repeat
>>> from pandas import *
>>>
>>> dummy = DataFrame({"foo": range(0, 5),
...                    "bar": range(10, 15),
...                    "b": [1, 2, 3, 5, 5],
...                    "a": list(repeat(3, 5)),
...                    "r": list(repeat(8, 5))})
>>>
>>> dummy.to_csv('dummy.csv', index=False)
>>> read_csv('dummy.csv', usecols=['foo'])
   foo
0    0
1    1
2    2
3    3
4    4
>>> read_csv('dummy.csv', usecols='bar')
   b  a  r
0  1  3  8
1  2  3  8
2  3  3  8
3  5  3  8
4  5  3  8
>>>
>>> os.remove('dummy.csv')
```

Added is_list_like check in _validate_usecols_arg to ensure passed value is list-like but not string. Also, array_like is defined as list_like with dtype attribute so update docstring to list-like from array-like.

### CODE REVIEW

**Comment by jreback** [comment]
can you add a whatsnew note.

@gfyoung can you review

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -97,11 +97,11 @@
     MultiIndex is used. If you have a malformed file with delimiters at the end
     of each line, you might consider index_col=False to force pandas to _not_
     use the first column as the index (row names)
-usecols : array-like or callable, default None
-    Return a subset of the columns. If array-like, all elements must either
+usecols : list-like or callable, default None
```
Comment (gfyoung) [inline pandas/io/parsers.py]: Leave this alone actually.  `array-like` will suffice.

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -445,10 +445,8 @@ cdef class TextReader:
         # suboptimal
         if usecols is not None:
             self.has_usecols = 1
-            if callable(usecols):
-                self.usecols = usecols
-            else:
-                self.usecols = set(usecols)
+            # GH20529, validate usecols at higher level.
```
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Let's expand this comment to this:

~~~
# gh-20558: we will validate and clean "usecols" before getting here.
~~~

**Comment by gfyoung** [comment]
This patch looks pretty good!  Save a few comments and the `whatsnew`, this should be merge-able.

File: pandas/io/parsers.py (hunk 1 above)
Comment (minggli) [reply pandas/io/parsers.py]: sure. but https://github.com/pandas-dev/pandas/blob/0a003651a67854b9776c46469a18e20ffd61f141/pandas/core/dtypes/inference.py#L287 will evaluate ['foo' ,'bar'] to False. so it's really a list-like IMHO.

File: pandas/_libs/parsers.pyx (hunk 2 above)
Comment (minggli) [reply pandas/_libs/parsers.pyx]: done.

File: pandas/io/parsers.py (hunk 1 above)
Comment (gfyoung) [reply pandas/io/parsers.py]: Hmmm...that's a fair point.  Actually, leave that alone then.

Comment (gfyoung) [reply pandas/io/parsers.py]: But if you're going to change it here, make sure to do it in `io.rst` as well.

Comment (minggli) [reply pandas/io/parsers.py]: sure. it's done.

**Review by gfyoung — APPROVED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
where is the test for a single string?

**Comment by minggli** [comment]
@jreback added test.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @minggli keep em coming!

**Comment by minggli** [comment]
happy to contribute. thanks for reviewing @jreback

**Pull request merged**